torch/csrc/jit/runtime/profiling_record.cpp

```diff
@@ -144,35 +144,43 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
   auto pno = pn->addOutput();
   pn->ty_(attr::profiled_type, TensorType::get());
   pno->setType(TensorType::get());
-  std::function<void(Stack&)> shape_profiler = [this, pno](Stack& stack) {
+  std::function<void(Stack&)> shape_profiler = [this, pn, pno](Stack& stack) {
     int64_t frame_id = 0;
     pop(stack, frame_id);
     IValue v;
     pop(stack, v);
+    std::lock_guard<std::mutex> lock(this->mutex_);
+    if (profiling_count_ <= 0) {
+      return;
+    }
+
     if (v.isTensor()) {
-      std::lock_guard<std::mutex> lock(this->mutex_);
-      auto& profiled_types = profiled_types_per_frame_[frame_id];
       auto& t = v.toTensor();
-      if (t.defined()) {
-        auto pttp = tensorTypeInCurrentExecutionContext(t);
+      auto new_tensor_type =
+          (t.defined() ? tensorTypeInCurrentExecutionContext(t)
+                       : TensorType::get()->withUndefined());
+      GRAPH_DEBUG(
+          "In run ",
+          frame_id,
+          " annotating %",
+          pno->debugName(),
+          " with ",
+          *new_tensor_type);
+
+      if (pn->hasRun()) {
+        auto existing_tensor_type =
+            pn->ty(attr::profiled_type)->expect<TensorType>();
```

Comment:
You don't need the shared_ptr, so use expectRef and save a reference count bump
`const auto& existing_tensor_type = pn->ty(attr::profiled_type)->expectRef<TensorType>();`